pkg/server/plugin/upstreamca/spire/spire_node.go

```diff
@@ -0,0 +1,241 @@
+package spireplugin
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/ecdsa"
+	"errors"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/spiffe/spire/proto/api/node"
+	"github.com/spiffe/spire/pkg/common/grpcutil"
+	"fmt"
+	"crypto/elliptic"
+	"crypto/rand"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/spire/pkg/common/log"
+)
+
+func (m *spirePlugin) getIntermediateCA (ctx context.Context, config *Configuration) error {
+//	TBD :
+//	get node connection
+//	create csr
+//	send  csr with CA option enabled
+//	return the intermediate certificates (intermediate CA , intermediate key and trust bundle
+	fmt.Println("[plugin][upstreamca] get intermediate certificates using node api")
+	nodeClient, err := m.newNodeAPIClient(ctx, config)
+	if err != nil {
+		fmt.Println("error seen due to node client create error")
+		return err
+	}
+
+	fmt.Println("create csr request with spiffe id [%s]",m.spiffeID)
+	// spiffe id to be used the one retrieved from workload api
+	requestCSR, err := m.createCSRRequest(m.spiffeID)
+	if err != nil {
+		return err
+	}
+	csrs := [][]byte{}
+	csrs = append(csrs,requestCSR)
+	nodeRequest := node.FetchX509SVIDRequest{Csrs:csrs}
+
+	stream, err := nodeClient.FetchX509SVID(ctx)
+	if err != nil {
+		fmt.Println("error in fetch node stream [%s]")
+		return err
+	}
+	err = stream.Send(&nodeRequest)
+	if err != nil {
+		return err
+	}
+
+	stream.CloseSend()
+
+	var nodeResponse *node.FetchX509SVIDResponse
+	nodeResponse, err = stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	m.intermediateCA, m.bundle, err = m.getCertFromResponse(nodeResponse)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *spirePlugin) submitCSRUpstreamCA (ctx context.Context, config *Configuration, csr []byte) error {
```

Comment:
I think this is a duplicate of the above function `getIntermediateCA`?

Reply:
Removed the redundant codes.